just: fix cross-compilation

xokdvium opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Fixes #320569.

Get rid of calls to `cargo run` in a, which lead to cargo TARGET and CC
to be out of sync for cross-compilation.

Use emulator to build man pages, shell completions and source for the docs.

With this change I'm able to build `pkgsCross.aarch64-multiplatform.just` as well as `pkgsCross.aarch64-multiplatform.pkgsStatic.just` from `x86_64-linux`.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alyssais commented on pkgs/by-name/ju/just/package.nix:
```diff
@@ -87,16 +68,41 @@ rustPlatform.buildRustPackage rec {
+          src
+          cargoHash
+          ;
+        nativeBuildInputs = [ buildPackages.mdbook ];
```
Given this is a build dependency for a build dependency, it should come from pkgsBuildBuild. (But also if you just use "mdbook" splicing should do the right thing.)

alyssais reviewed (commented):
An easier and more portable approach would probably be to just copy the documentation from buildPackages.just when cross compiling.

xokdvium commented:
> An easier and more portable approach would probably be to just copy the documentation from buildPackages.just when cross compiling.

Thanks for the tip! I wasn't sure how to best handle it in this case. 
In fact the docs were building fine when retargeting cargo for the host system, but that seemed fragile to me. I implemented it how you suggested: 

For native compilation we buid the docs if `withDocs` is set, if we are cross-compiling then they get copied from `buildPackages.just`.

It's a bit more work to compile `just` natively than only the docs, but a more general solution is preferable in this case.

xokdvium commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 97 packages built:</summary>
  <ul>
    <li>aba</li>
    <li>catppuccin-cursors</li>
    <li>catppuccin-cursors.frappeBlue</li>
    <li>catppuccin-cursors.frappeDark</li>
    <li>catppuccin-cursors.frappeFlamingo</li>
    <li>catppuccin-cursors.frappeGreen</li>
    <li>catppuccin-cursors.frappeLavender</li>
    <li>catppuccin-cursors.frappeLight</li>
    <li>catppuccin-cursors.frappeMaroon</li>
    <li>catppuccin-cursors.frappeMauve</li>
    <li>catppuccin-cursors.frappePeach</li>
    <li>catppuccin-cursors.frappePink</li>
    <li>catppuccin-cursors.frappeRed</li>
    <li>catppuccin-cursors.frappeRosewater</li>
    <li>catppuccin-cursors.frappeSapphire</li>
    <li>catppuccin-cursors.frappeSky</li>
    <li>catppuccin-cursors.frappeTeal</li>
    <li>catppuccin-cursors.frappeYellow</li>
    <li>catppuccin-cursors.latteBlue</li>
    <li>catppuccin-cursors.latteDark</li>
    <li>catppuccin-cursors.latteFlamingo</li>
    <li>catppuccin-cursors.latteGreen</li>
    <li>catppuccin-cursors.latteLavender</li>
    <li>catppuccin-cursors.latteLight</li>
    <li>catppuccin-cursors.latteMaroon</li>
    <li>catppuccin-cursors.latteMauve</li>
    <li>catppuccin-cursors.lattePeach</li>
    <li>catppuccin-cursors.lattePink</li>
    <li>catppuccin-cursors.latteRed</li>
    <li>catppuccin-cursors.latteRosewater</li>
    <li>catppuccin-cursors.latteSapphire</li>
    <li>catppuccin-cursors.latteSky</li>
    <li>catppuccin-cursors.latteTeal</li>
    <li>catppuccin-cursors.latteYellow</li>
    <li>catppuccin-cursors.macchiatoBlue</li>
    <li>catppuccin-cursors.macchiatoDark</li>
    <li>catppuccin-cursors.macchiatoFlamingo</li>
    <li>catppuccin-cursors.macchiatoGreen</li>
    <li>catppuccin-cursors.macchiatoLavender</li>
    <li>catppuccin-cursors.macchiatoLight</li>
    <li>catppuccin-cursors.macchiatoMaroon</li>
    <li>catppuccin-cursors.macchiatoMauve</li>
    <li>catppuccin-cursors.macchiatoPeach</li>
    <li>catppuccin-cursors.macchiatoPink</li>
    <li>catppuccin-cursors.macchiatoRed</li>
    <li>catppuccin-cursors.macchiatoRosewater</li>
    <li>catppuccin-cursors.macchiatoSapphire</li>
    <li>catppuccin-cursors.macchiatoSky</li>
    <li>catppuccin-cursors.macchiatoTeal</li>
    <li>catppuccin-cursors.macchiatoYellow</li>
    <li>catppuccin-cursors.mochaBlue</li>
    <li>catppuccin-cursors.mochaDark</li>
    <li>catppuccin-cursors.mochaFlamingo</li>
    <li>catppuccin-cursors.mochaGreen</li>
    <li>catppuccin-cursors.mochaLavender</li>
    <li>catppuccin-cursors.mochaLight</li>
    <li>catppuccin-cursors.mochaMaroon</li>
    <li>catppuccin-cursors.mochaMauve</li>
    <li>catppuccin-cursors.mochaPeach</li>
    <li>catppuccin-cursors.mochaPink</li>
    <li>catppuccin-cursors.mochaRed</li>
    <li>catppuccin-cursors.mochaRosewater</li>
    <li>catppuccin-cursors.mochaSapphire</li>
    <li>catppuccin-cursors.mochaSky</li>
    <li>catppuccin-cursors.mochaTeal</li>
    <li>catppuccin-cursors.mochaYellow</li>
    <li>catppuccin-sddm</li>
    <li>celeste</li>
    <li>cosmic-applets</li>
    <li>cosmic-applibrary</li>
    <li>cosmic-bg</li>
    <li>cosmic-design-demo</li>
    <li>cosmic-edit</li>
    <li>cosmic-files</li>
    <li>cosmic-greeter</li>
    <li>cosmic-icons</li>
    <li>cosmic-launcher</li>
    <li>cosmic-notifications</li>
    <li>cosmic-panel</li>
    <li>cosmic-randr</li>
    <li>cosmic-screenshot</li>
    <li>cosmic-session</li>
    <li>cosmic-settings</li>
    <li>cosmic-store</li>
    <li>cosmic-term</li>
    <li>dogdns</li>
    <li>dogdns.man</li>
    <li>find-billy</li>
    <li>forecast</li>
    <li>just</li>
    <li>just.doc</li>
    <li>just.man</li>
    <li>kabeljau</li>
    <li>onagre</li>
    <li>pop-launcher</li>
    <li>ssh-openpgp-auth</li>
    <li>sshd-openpgp-auth</li>
  </ul>
</details>

xokdvium commented:
Also able to build `pkgsCross.riscv64.just` in addition to the dynamic/static builds for `aarch64`  from `x86_64-linux`

alyssais commented on pkgs/by-name/ju/just/package.nix:
```diff
@@ -87,16 +83,31 @@ rustPlatform.buildRustPackage rec {
+    )
+    + lib.optionalString (stdenv.hostPlatform.emulatorAvailable buildPackages) ''
+      ${just} --man > ./just.1
+      installManPage ./just.1
```
Why aren't we copying this from buildPackages.just as well?

xokdvium replied:
I've looked back on the solution with the copying and it doesn't really sit well with me. 
I refactored `generate-book` to use emulation. As it turns out it's not a bottleneck at all and takes no time at all.
Sorry for the back and forth

alyssais replied:
Emulation should really only be used where it's a last resort, because we don't have emulators for every platform we might want to cross compile to.

Aleksanaa replied:
We've discussed it a bit here https://github.com/NixOS/nixpkgs/pull/351032, but I don't have a clear opinion on this

alyssais replied:
Wrong link?

Aleksanaa replied:
Oh https://github.com/NixOS/nixpkgs/issues/308283

xokdvium committed:
just: fix cross-compilation

Get rid of calls to `cargo run` in a, which lead to cargo TARGET and CC
to be out of sync for cross-compilation.

Only build man pages, shell completions and source for the docs when
`stdenv.buildPlatform.canExecute stdenv.hostPlatform`, as this seems to be
the current consensus.

06kellyjac commented on pkgs/by-name/ju/just/package.nix:
```diff
@@ -9,16 +9,22 @@
   libiconv,
   mdbook,
   nix-update-script,
+  installShellCompletions ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
+  installManPages ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
+  withDocumentation ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
```
```suggestion
  # run the compiled `just` to build the completions
  installShellCompletions ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
  # run the compiled `just` to build the man pages
  installManPages ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
  # run the compiled `generate-book` utility to prepare the files for mdbook
  withDocumentation ? stdenv.buildPlatform.canExecute stdenv.hostPlatform,
```

Just proposing some comments as initially someone (such as myself :sweat_smile: ) might think `mdbook` and `installManPage` are going to be executable on `buildPlatform` coming from `nativeBuildInputs` but it's more about using `just` or `generate-book`

06kellyjac reviewed (approved):
Non-blocking suggestion to help any future reviewers of the code.

I'm happy with this method or copying results from `buildPackages`. Emulation methods are fun but as discussed in the other issue require an excessively large closure and can be brittle.

xokdvium committed:
just: add comments for completion and doc generation

Co-authored-by: j-k <dev@j-k.io>

misuzu commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 351032`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 98 packages built:</summary>
  <ul>
    <li>aba</li>
    <li>catppuccin-cursors</li>
    <li>catppuccin-cursors.frappeBlue</li>
    <li>catppuccin-cursors.frappeDark</li>
    <li>catppuccin-cursors.frappeFlamingo</li>
    <li>catppuccin-cursors.frappeGreen</li>
    <li>catppuccin-cursors.frappeLavender</li>
    <li>catppuccin-cursors.frappeLight</li>
    <li>catppuccin-cursors.frappeMaroon</li>
    <li>catppuccin-cursors.frappeMauve</li>
    <li>catppuccin-cursors.frappePeach</li>
    <li>catppuccin-cursors.frappePink</li>
    <li>catppuccin-cursors.frappeRed</li>
    <li>catppuccin-cursors.frappeRosewater</li>
    <li>catppuccin-cursors.frappeSapphire</li>
    <li>catppuccin-cursors.frappeSky</li>
    <li>catppuccin-cursors.frappeTeal</li>
    <li>catppuccin-cursors.frappeYellow</li>
    <li>catppuccin-cursors.latteBlue</li>
    <li>catppuccin-cursors.latteDark</li>
    <li>catppuccin-cursors.latteFlamingo</li>
    <li>catppuccin-cursors.latteGreen</li>
    <li>catppuccin-cursors.latteLavender</li>
    <li>catppuccin-cursors.latteLight</li>
    <li>catppuccin-cursors.latteMaroon</li>
    <li>catppuccin-cursors.latteMauve</li>
    <li>catppuccin-cursors.lattePeach</li>
    <li>catppuccin-cursors.lattePink</li>
    <li>catppuccin-cursors.latteRed</li>
    <li>catppuccin-cursors.latteRosewater</li>
    <li>catppuccin-cursors.latteSapphire</li>
    <li>catppuccin-cursors.latteSky</li>
    <li>catppuccin-cursors.latteTeal</li>
    <li>catppuccin-cursors.latteYellow</li>
    <li>catppuccin-cursors.macchiatoBlue</li>
    <li>catppuccin-cursors.macchiatoDark</li>
    <li>catppuccin-cursors.macchiatoFlamingo</li>
    <li>catppuccin-cursors.macchiatoGreen</li>
    <li>catppuccin-cursors.macchiatoLavender</li>
    <li>catppuccin-cursors.macchiatoLight</li>
    <li>catppuccin-cursors.macchiatoMaroon</li>
    <li>catppuccin-cursors.macchiatoMauve</li>
    <li>catppuccin-cursors.macchiatoPeach</li>
    <li>catppuccin-cursors.macchiatoPink</li>
    <li>catppuccin-cursors.macchiatoRed</li>
    <li>catppuccin-cursors.macchiatoRosewater</li>
    <li>catppuccin-cursors.macchiatoSapphire</li>
    <li>catppuccin-cursors.macchiatoSky</li>
    <li>catppuccin-cursors.macchiatoTeal</li>
    <li>catppuccin-cursors.macchiatoYellow</li>
    <li>catppuccin-cursors.mochaBlue</li>
    <li>catppuccin-cursors.mochaDark</li>
    <li>catppuccin-cursors.mochaFlamingo</li>
    <li>catppuccin-cursors.mochaGreen</li>
    <li>catppuccin-cursors.mochaLavender</li>
    <li>catppuccin-cursors.mochaLight</li>
    <li>catppuccin-cursors.mochaMaroon</li>
    <li>catppuccin-cursors.mochaMauve</li>
    <li>catppuccin-cursors.mochaPeach</li>
    <li>catppuccin-cursors.mochaPink</li>
    <li>catppuccin-cursors.mochaRed</li>
    <li>catppuccin-cursors.mochaRosewater</li>
    <li>catppuccin-cursors.mochaSapphire</li>
    <li>catppuccin-cursors.mochaSky</li>
    <li>catppuccin-cursors.mochaTeal</li>
    <li>catppuccin-cursors.mochaYellow</li>
    <li>catppuccin-sddm</li>
    <li>celeste</li>
    <li>cosmic-applets</li>
    <li>cosmic-applibrary</li>
    <li>cosmic-bg</li>
    <li>cosmic-design-demo</li>
    <li>cosmic-edit</li>
    <li>cosmic-files</li>
    <li>cosmic-greeter</li>
    <li>cosmic-icons</li>
    <li>cosmic-launcher</li>
    <li>cosmic-notifications</li>
    <li>cosmic-panel</li>
    <li>cosmic-randr</li>
    <li>cosmic-screenshot</li>
    <li>cosmic-session</li>
    <li>cosmic-settings</li>
    <li>cosmic-store</li>
    <li>cosmic-term</li>
    <li>dogdns</li>
    <li>dogdns.man</li>
    <li>find-billy</li>
    <li>forecast</li>
    <li>just</li>
    <li>just.doc</li>
    <li>just.man</li>
    <li>kabeljau</li>
    <li>onagre</li>
    <li>pop-launcher</li>
    <li>rofi-games</li>
    <li>ssh-openpgp-auth</li>
    <li>sshd-openpgp-auth</li>
  </ul>
</details>

misuzu commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 351032`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 10 packages built:</summary>
  <ul>
    <li>aba</li>
    <li>celeste</li>
    <li>dogdns</li>
    <li>dogdns.man</li>
    <li>just</li>
    <li>just.doc</li>
    <li>just.man</li>
    <li>kabeljau</li>
    <li>ssh-openpgp-auth</li>
    <li>sshd-openpgp-auth</li>
  </ul>
</details>
